File: Week 02/id_037/课堂习题/generate-parentheses.java
// https://leetcode-cn.com/problems/generate-parentheses/
class Solution {
    
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList();
        int left = 0;
        int right = 0;
        String curr = "";
        helper(n, left, right, curr, res);
        return res;
    }
    private void helper(int n, int left, int right, String curr, List<String> res) {
        if (left + right == 2 * n) {
            res.add(curr);
            return;
        }
        if (left < n) {
            helper(n, left++, right, curr + "(", res);
        }
        if (right < left) {
            helper(n, left, right++, curr+")", res);
        }
    }
}

File: Week 06/id_262/generate-parentheses.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def _helper(s, left, right):
            if len(s) == 2 * n:
                res.append(s)
            if left < n:
                _helper(s + "(", left + 1, right)
            if right < left:
                _helper(s + ")", left, right + 1)

        res = []
        _helper("", 0, 0)
        return res
